caffe2/core/blob.h

```diff
@@ -24,6 +25,41 @@ inline bool BlobIsTensorType(const Blob& blob, DeviceType device_type) {
   return tensor && *tensor && tensor->GetDeviceType() == device_type;
 }
 
+inline void BlobSetTensor(Blob* blob, Tensor tensor) {
+  blob->Reset<Tensor>(new Tensor(std::move(tensor)));
+}
+
+inline Tensor* BlobGetMutableTensor(
+    Blob* blob,
+    const vector<int64_t>& dims,
+    const at::TensorOptions& options) {
+  if (blob->IsType<Tensor>()) {
+    Tensor* tensor = blob->GetMutable<Tensor>();
+    if (*tensor) {
+      if (tensor->GetDeviceType() ==
+          at::backendToDeviceType(options.backend())) {
+        if (tensor->dims() != dims) {
+          // Resize when the dims doesn't match
+          tensor->Resize(dims);
+        } else {
+          // reallocate when the data_type doesn't match
+          if (tensor->meta() != at::scalarTypeToTypeMeta(options.dtype())) {
+            tensor->raw_mutable_data(at::scalarTypeToTypeMeta(options.dtype()));
+          }
+        }
+        return tensor;
+      }
+      // If DeviceType is wrong, we'll allocate a new Tensor
+    }
+  }
+
+  // if we're here, then either Blob didn't hold a Tensor
+  // or that Tensor had the wrong DeviceType.
+  VLOG(1) << "Create new mutable object " << TypeMeta::TypeName<Tensor>()
+          << " dims: " << dims << " options: " << options;
+  return blob->Reset<Tensor>(new Tensor(empty(dims, options)));
```

Comment:
Should we not do empty but also explicitly create the storage via raw_mutable_data? If we now have the dims and data type, it seems natural that we can allocate the underlying memory now.

Reply:
what do you mean? empty will call raw_mutable_data to allocate memory.

Reply:
PyTorch could always add a `torch.uninitialized` function, which means the same thing as empty. It's not very Pythonic, but it seems pretty clear.